p-ata/src/processor.rs

```diff
@@ -0,0 +1,982 @@
+                        bump
+                    );
+                    return Err(ProgramError::InvalidSeeds);
+                }
+            }
+            None => {
+                let (canonical_address, _bump) = derive_canonical_ata_pda(
+                    wallet.key(),
+                    token_program.key(),
+                    mint_account.key(),
+                    program_id,
+                );
+
+                if canonical_address != *associated_token_account.key() {
+                    log!(
+                        "Error: Provided associated token account address is not canonical. Expected: {}, Found: {}",
+                        &canonical_address,
+                        associated_token_account.key()
+                    );
+                    return Err(ProgramError::InvalidSeeds);
+                }
+            }
+        }
+
+        return Ok(true);
+    }
+    Ok(false)
+}
+
+/// Determine the required space (in bytes) for the associated token account.
+///
+/// Either uses a pre-computed account length hint or calls into the size
+/// calculation logic to determine the space needed. For extended accounts,
+/// passing in length can save significant compute units.
+///
+/// ## Arguments
+///
+/// * `known_token_account_len` - Optional pre-computed account length
+///
+/// ## Returns
+///
+/// The account size in bytes, or an error if size calculation fails.
+#[inline(always)]
+pub(crate) fn resolve_token_account_space(
+    token_program: &AccountInfo,
+    mint_account: &AccountInfo,
+    known_token_account_len: Option<usize>,
+) -> Result<usize, ProgramError> {
+    match known_token_account_len {
+        Some(len) => Ok(len),
+        None => get_token_account_size(mint_account, token_program),
+    }
+}
+
+/// Create and initialize an ATA account with the given bump seed.
+#[allow(clippy::too_many_arguments)]
+#[inline(always)]
+pub(crate) fn create_and_initialize_ata(
+    payer: &AccountInfo,
+    associated_token_account: &AccountInfo,
+    wallet: &AccountInfo,
+    mint_account: &AccountInfo,
+    token_program: &AccountInfo,
+    rent: &Rent,
+    bump: u8,
+    space: usize,
+) -> ProgramResult {
+    let bump_slice = [bump];
+    let seeds: [Seed; 4] = [
+        Seed::from(wallet.key().as_ref()),
+        Seed::from(token_program.key().as_ref()),
+        Seed::from(mint_account.key().as_ref()),
+        Seed::from(&bump_slice[..]),
+    ];
+
+    create_pda_account(
+        payer,
+        rent,
+        space,
+        token_program.key(),
+        associated_token_account,
+        &seeds,
+    )?;
+
+    // Initialize ImmutableOwner for non-SPL Token programs (future compatible)
+    if !is_spl_token_program(token_program.key()) {
+        let initialize_immutable_owner_metas = &[AccountMeta {
+            pubkey: associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        }];
```

Comment:
`AccountMeta` has some helper constructors:
```suggestion
        let initialize_immutable_owner_metas =
            &[AccountMeta::writable(associated_token_account.key())];
```